cmds/id/id.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2013-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+// Print process information.
+//
+// Synopsis:
+//      id [-gGnu]
+//
+// Description:
+//      id displays the uid, guid and groups of the calling process
+// Options:
+//  		-g, --group     print only the effective group ID
+//		  -G, --groups    print all group IDs
+//		  -n, --name      print a name instead of a number, for -ugG
+//		  -u, --user      print only the effective user ID
+//
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+var (
+	GROUP_FILE  = "/etc/group"
+	PASSWD_FILE = "/etc/passwd"
+	l           = log.New(os.Stderr, "", 0)
+
+	flags struct {
+		g bool
+		G bool
+		n bool
+		u bool
+	}
+)
+
+func correctFlags(flags ...bool) bool {
+	n := 0
+	for _, v := range flags {
+		if v {
+			n += 1
+		}
+	}
+	if n > 1 {
+		return false
+	} else {
+		return true
+	}
+}
+
+func initFlags() error {
+	flag.BoolVar(&flags.g, "g", false, "print only the effective group ID")
+	flag.BoolVar(&flags.G, "G", false, "print all group IDs")
+	flag.BoolVar(&flags.n, "n", false, "print a name instead of a number, for -ugG")
+	flag.BoolVar(&flags.u, "u", false, "print only the effective user ID")
+	flag.Parse()
+	if !correctFlags(flags.G, flags.g, flags.u) {
+		return fmt.Errorf("cannot print \"only\" of more than one choice\n")
+
+	}
+	if flags.n && !(flags.G || flags.g || flags.u) {
+		return fmt.Errorf("cannot print only names in default format\n")
+	}
+
+	return nil
+}
+
+type User struct {
```

Comment:
Alright, so, here's the thing about this struct. Usually, if you are going to have a struct like this, you don't have a bunch of pull methods, but a "constructor" (Go doesn't have constructors, but you can have one anyway):

```go
func NewUser() (*User, error) {
  u := &User{
    UID: syscall.Getuid(),
    EUID: syscall.Geteuid(),
    GID: syscall.Getgid(),
  }
  // get name and fill in u.Name
 
  // get groups, fill in u.Groups
  return u, nil
}
```

The other thing is that in Go, methods never start with `Get`. It's just `user.UID()`, the `get` is superfluous to Go programmers... 

This also means you never have a *User with incomplete information, which would be easier to do if you forgot to call "pullUid" on some user.